File: 3981-jump-game-ix/jump-game-ix.py
from typing import List

class Solution:
    def maxValue(self, nums: List[int]) -> List[int]:
        n = len(nums)

        # suffix minimum array
        sufMin = [float('inf')] * (n + 1)

        for i in range(n - 1, -1, -1):
            sufMin[i] = min(sufMin[i + 1], nums[i])

        ans = [0] * n

        l = 0
        mx = nums[0]

        for r in range(n):
            mx = max(mx, nums[r])

            # end of connected component
            if r == n - 1 or mx <= sufMin[r + 1]:
                for i in range(l, r + 1):
                    ans[i] = mx

                l = r + 1

                if l < n:
                    mx = nums[l]

        return ans

File: 3981-jump-game-ix/jump-game-ix.cpp
class Solution {
public:
    vector<int> maxValue(vector<int>& nums) {
        int n = (int)nums.size();
        vector<int> sufMin(n + 1, INT_MAX);

        // sufMin[i] = minimum value in nums[i...n-1]
        for (int i = n - 1; i >= 0; --i) {
            sufMin[i] = min(sufMin[i + 1], nums[i]);
        }

        vector<int> ans(n);
        int l = 0;          // start of current component
        int mx = nums[0];   // max inside current component

        for (int r = 0; r < n; ++r) {
            mx = max(mx, nums[r]);

            // If no inversion crosses the cut after r, close this component
            if (r == n - 1 || mx <= sufMin[r + 1]) {
                for (int i = l; i <= r; ++i) {
                    ans[i] = mx;
                }
                l = r + 1;
                if (l < n) mx = nums[l];
            }
        }

        return ans;
    }
};